### ISSUE TITLE
CA1859 erroneously suggests changing return type to one of multiple possible return types

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn-analyzers#6411]
### Analyzer

**Diagnostic ID**: CA1859

### Analyzer source

https://github.com/dotnet/roslyn-analyzers/commit/6ae2683c5033dfde0b73d4e73b372f00ca6ceacd

### Describe the bug

A method can return both the base abstract type and a derived type, and the analyzer erroneously suggests changing the return type to the derived type.

### Steps To Reproduce

```C#
[Fact]
public static async Task Repro()
{
    const string Source = @"
        #nullable enable

        using System;
        using System.IO;

        namespace Example
        {
            internal static class C
            {
                private static Stream GetStream(int i)
                {
                    if (i == 0)
                    {
                        return Stream.Null;
                    }

                    return new MyStream();
                }
            }
        }

        public class MyStream : MemoryStream { }
        ";

    await TestCSAsync(Source);
}
```

### Expected behavior

No warning.

### Actual behavior

```
    // /0/Test0.cs(11,47): info CA1859: Change return type of method 'GetStream' from 'System.IO.Stream' to 'MyStream' for improved performance
VerifyCS.Diagnostic(UseConcreteTypeAnalyzer.CA1859).WithSpan(11, 47, 11, 56).WithArguments("GetStream", "System.IO.Stream", "MyStream"),
```

cc: @geeknoid, @buyaa-n

**Comment by Youssef1313** [comment]
Closing as fixed per https://github.com/dotnet/roslyn-analyzers/pull/6418#issuecomment-1385520293
